Add support for inline image embeding

carliros opened the pull request:
Resolves https://github.com/srid/emanote/issues/98

carliros committed:
Add support for inline image embeding

https://github.com/srid/emanote/issues/98

carliros commented:
Here is a screenshot: 

![Screen Shot 2021-07-16 at 14 11 36](https://user-images.githubusercontent.com/151595/125991162-210e780a-0451-4d2c-a11c-6442736a1386.png)

srid commented on default/templates/filters/embed-image.tpl:
```diff
@@ -1 +1 @@
-<img class="mb-3" src="${ema:url}" alt="${ema:alt}" />
\ No newline at end of file
+<img class="mb-3" src="${ema:url}" alt="${ema:alt}" style="display: inline;" />
```
You can replace this with Tailwind's [`inline`](https://tailwindcss.com/docs/display#inline) class. 

More generally we should use Tailwind's classes wherever applicable, and resort to `style` attribute only when absolutely necessary. Tailwind is a feature of the default templates (you can use them index.yaml for instance).

srid commented on src/Emanote/Pandoc/Renderer/Embed.hs:
```diff
@@ -43,14 +43,41 @@ embedWikiLinkResolvingSplice emaAction model nf (ctxSansCustomSplicing -> ctx) _
+  Monad n => PandocInlineRenderer n x
+embedInlineWikiLinkResolvingSplice emaAction model nf (ctxSansCustomSplicing -> ctx) _ inl =
+  case inl of
+    B.Link attr@(_id, _class, otherAttrs) is (url, tit) -> do
```
You can use `LambdaCase` here instead of using `inl`?

srid commented on src/Emanote/Pandoc/Renderer/Embed.hs:
```diff
@@ -43,14 +43,41 @@ embedWikiLinkResolvingSplice emaAction model nf (ctxSansCustomSplicing -> ctx) _
+      HE.runCustomTemplate tpl splices
+
+embedInlineSiteRoute :: Monad n => Ema.CLI.Action -> Model -> NoteRenderers n -> HP.RenderCtx n -> WL.WikiLink -> Either MN.Note SF.StaticFile -> Maybe (HI.Splice n)
+embedInlineSiteRoute _ _ _ _ wl = \case
```
As a general Haskell coding guideline, if an argument is never used in a function, it should be removed from its signature. These 4 arguments should be removed. 

Why? The functional signature should be as simple as possible for understanding. The signature in part acts as its 'documentation', in that - reading the signature more or less tells you what this function does.

srid commented on src/Emanote/Pandoc/Renderer/Embed.hs:
```diff
@@ -43,14 +43,41 @@ embedWikiLinkResolvingSplice emaAction model nf (ctxSansCustomSplicing -> ctx) _
+
+embedInlineSiteRoute :: Monad n => Ema.CLI.Action -> Model -> NoteRenderers n -> HP.RenderCtx n -> WL.WikiLink -> Either MN.Note SF.StaticFile -> Maybe (HI.Splice n)
+embedInlineSiteRoute _ _ _ _ wl = \case
+  Left _ -> Nothing
```
Could you add comment stating *why* this case is `Nothing`?

EDIT: see other comment.

carliros replied:
being addressed with the refactoring suggested in other comment.

srid commented on src/Emanote/View/Template.hs:
```diff
@@ -93,8 +93,10 @@ renderLmlHtml emaAction model note = do
-          [PF.urlResolvingSplice]
-          [ PF.embedWikiLinkResolvingSplice,
+          [ PF.embedInlineWikiLinkResolvingSplice,
+            PF.urlResolvingSplice
```
Is there a reason for changing the order of `urlResolvingSplice` (it is no longer the first spice in this PR)?

carliros replied:
This was because having `urlResolvingSplice` first does not recognize the inline element (Link) as image but as File embed but not rendered as image. Putting first `PF.embedInlineWikiLinkResolvingSplice` fixed that.

srid replied:
In that case, we should add a code comment saying exactly that (for posterity).

srid commented on docs/demo/embed.md:
```diff
@@ -28,6 +28,8 @@ Embedding image files as, say, `![[disaster-girl.jpg]]` is equivalent to `![](pa
 
 ![[disaster-girl.jpg]]
 
+It is also posible to add ![[ema-small.png]] as inline paragraphs.
```
Do we really need to add a new image file to the repository? Can't we use `favicon.svg`? Shouldn't something like this work?

```markdown
It is also posible to add images inline (example, here's the 
site favicon: [![[favicon.svg]]]{.w-6}) say in the middle of a paragraph.
```

carliros replied:
Good to know that we can use that image as well.

srid commented on src/Emanote/Pandoc/Renderer/Embed.hs:
```diff
@@ -43,14 +43,41 @@ embedWikiLinkResolvingSplice emaAction model nf (ctxSansCustomSplicing -> ctx) _
       "ema:note:url" ## HI.textSplice (SR.siteRouteUrl model $ SR.lmlSiteRoute $ note ^. MN.noteRoute)
       "ema:note:pandoc"
         ## noteSpliceWith nf classMap emaAction model note
+  Right _ -> Nothing
```
Could you add comment stating why this case is Nothing?

EDIT: see other comment.

carliros replied:
being addressed with the refactoring suggested in other comment.

srid commented on src/Emanote/Pandoc/Renderer/Embed.hs:
```diff
@@ -43,14 +43,41 @@ embedWikiLinkResolvingSplice emaAction model nf (ctxSansCustomSplicing -> ctx) _
         ## noteSpliceWith nf classMap emaAction model note
+  Right _ -> Nothing
+  where
+    runEmbedTemplate name splices = do
```
It looks like `runEmbedTemplate ` is used in the other function as well? If so it should be made a top-level, an reused without duplication.

srid commented on src/Emanote/Pandoc/Renderer/Embed.hs:
```diff
@@ -43,14 +43,41 @@ embedWikiLinkResolvingSplice emaAction model nf (ctxSansCustomSplicing -> ctx) _
+        Left err ->
+          pure $ brokenLinkSpanWrapper err $ B.Link attr (Url.nonEmptyLinkInlines model url Nothing is) (url, tit)
+        Right res -> do
+          embedInlineSiteRoute emaAction model nf ctx wl res
```
Okay, as a follow-up to two other comments on `Nothing` further below, I think it would be better to refactor this into two functions. One that takes `noteRoute` as argument and returns a block embedding splice (call it `embedResourceRoute`); another that takes `staticFileRoute` as argument and returns an inline embedding splice (call it `embedStaticFileRoute`). 

Then you can use it here as,

```haskell
either (embedResourceNote emaAction model nf ctx wl) (const Nothing) res
```

and in the inline splice,

```haskell
either (const Nothing) (embedStaticFileRoute emaAction model nf ctx wl) res
```

srid reviewed (changes requested):
Thanks! Added my feedback.

srid commented:
Also .cabal file version should be incremented.

carliros commented:
Super cool, thanks for the feedback, I will address each of them. I am learning a lot on participating on this project.

srid commented:
@carliros #109 refactored the renderer types; you will want to merge from `master` to resolve the conflicts. Let me know if the conflict resolution is confusing, I can help.

EDIT: It might be better to re-do your changes on top of current `master`

carliros committed:
update version

carliros commented:
@srid I am done with the refactoring. Please take a look.

carliros commented:
One think might need some extra explanation is the add of `width: inherit` in
```
<img class="inline mb-3" src="${ema:url}" alt="${ema:alt}" style="width: inherit;" />
```

That is needed in order to respect the father width with the child element. Without that, the image is rendered in full width.

srid committed:
Fix hang when using broken inline image

Use `ctxSansCustomSplicing` to avoid recursing into custom splicing.

srid committed:
Introduce BrokenLink type (refactor)

This also fixes the case where using "![[Foo.jpg]]" on its own paragraph
leads to empty rendering.

srid committed:
Make BrokenLink its own module

srid committed:
Make Url.hs to (re)use the BrokenLink type

srid committed:
Refactor: use ctxSansCustomSplicing in one place only

srid commented:
Just refactored a few things, mainly the "broken link" handling/rendering was spread across Url.hs and Embed.hs; I've consolidated it into one place, so in future it will be simpler to reason about this concept from multiple places. As a meta/pedagogic note, it is quite common to do a "refactor postmortem" on feature PRs like this, so as to make sure that avoidable complexity can be kept at bay (as the code size grows).

EDIT: btw, this refactor also uncovered a hang bug, as fixed in the first commit.
